util/wait_group_wrapper.go

```diff
@@ -16,8 +16,121 @@ package util
 
 import (
 	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
 )
 
+// WaitGroupEnhancedWrapper wrapper wg
+type WaitGroupEnhancedWrapper struct {
+	sync.WaitGroup
+	exited          *atomic.Bool
+	source          string
+	registerProcess sync.Map
+}
+
+// NewWaitGroupEnhancedWrapper returns WaitGroupEnhancedWrapper
+func NewWaitGroupEnhancedWrapper(source string, exited *atomic.Bool) *WaitGroupEnhancedWrapper {
+	wgew := &WaitGroupEnhancedWrapper{
+		exited:          exited,
+		source:          source,
+		registerProcess: sync.Map{},
+	}
+	if len(source) > 0 {
+		go wgew.checkUnExitedProcess()
+	}
+	return wgew
+}
+
+func (w *WaitGroupEnhancedWrapper) checkUnExitedProcess() {
+	logutil.BgLogger().Info("waitGroupWrapper ready to check unexited process", zap.String("source", w.source))
+	ticker := time.NewTimer(10 * time.Second)
+	defer ticker.Stop()
+	for {
+		<-ticker.C
+		continueCheck := w.check()
+		if !continueCheck {
+			return
+		}
+	}
+}
+
+func (w *WaitGroupEnhancedWrapper) check() bool {
+	if !w.exited.Load() {
+		return true
+	}
+	unexitedProcess := make([]string, 0)
+	w.registerProcess.Range(func(key, value any) bool {
+		unexitedProcess = append(unexitedProcess, key.(string))
+		return true
+	})
+	if len(unexitedProcess) > 0 {
+		logutil.BgLogger().Warn("background process unexited while received exited signal",
+			zap.Strings("process", unexitedProcess),
+			zap.String("source", w.source))
+		return true
+	}
+	logutil.BgLogger().Info("waitGroupWrapper finish checking unexited process", zap.String("source", w.source))
+	return false
+}
+
+// Run runs a function in a goroutine, adds 1 to WaitGroup
+// and calls done when function returns. Please DO NOT use panic
+// in the cb function.
+func (w *WaitGroupEnhancedWrapper) Run(exec func(), label string) {
+	w.onStart(label)
+	w.Add(1)
+	go func() {
+		defer func() {
+			w.Done()
```

Comment:
seems when the caller use w.Wait() then set `exited` to true, w.onExit() may not be executed so the `label` is not deleted in w.registerProcess. It will be deleted soon but I'm not sure if this false-alert is critical

Reply:
updated.